experimenter/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/FormFeatureValue/RichFeatureValueEditor.tsx

```diff
@@ -0,0 +1,413 @@
+/* istanbul ignore file */
+
+import { SchemaDraft, validate } from "@cfworker/json-schema";
+import { autocompletion, CompletionContext } from "@codemirror/autocomplete";
+import { json } from "@codemirror/lang-json";
+import { syntaxTree } from "@codemirror/language";
+import { Diagnostic, linter } from "@codemirror/lint";
+import { EditorState, Text } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import ReactCodeMirror, { ReactCodeMirrorRef } from "@uiw/react-codemirror";
+import classNames from "classnames";
+import jsonToAst from "json-to-ast";
+import React, { useMemo, useRef } from "react";
+import { ControllerRenderProps } from "react-hook-form";
+import { useCommonNestedForm } from "src/hooks";
+import { getConfig_nimbusConfig_allFeatureConfigs } from "src/types/getConfig";
+import { z } from "zod";
+
+export interface RichFeatureValueEditorProps {
+  featureConfig: getConfig_nimbusConfig_allFeatureConfigs;
+  field: ControllerRenderProps<Record<string, any>>;
+  formControlAttrs: ReturnType<
+    ReturnType<typeof useCommonNestedForm>["formControlAttrs"]
+  >;
+}
+
+export default function RichFeatureValueEditor({
+  featureConfig,
+  field,
+  formControlAttrs,
+}: RichFeatureValueEditorProps) {
+  const editor = useRef<ReactCodeMirrorRef>(null);
+
+  const schema = useMemo<Record<string, unknown> | null>(() => {
+    if (!featureConfig.schema) {
+      return null;
+    }
+
+    try {
+      return JSON.parse(featureConfig.schema);
+    } catch (e) {
+      return null;
+    }
+  }, [featureConfig.schema]);
+
+  const { onChange: hideSubmitErrors } = formControlAttrs;
+
+  const extensions = useMemo(() => {
+    const extensions = [
+      json(),
+      EditorView.focusChangeEffect.of(
+        (state: EditorState, focusing: boolean) => {
+          if (!focusing) {
+            const value = state.doc.toString();
+
+            // We cannot trigger a change while a React update is on the stack.
+            setTimeout(() => {
+              field.onChange(value);
+            }, 0);
+
+            hideSubmitErrors?.();
+          }
+
+          return null;
+        },
+      ),
+    ];
+
+    if (schema) {
+      extensions.push(linter(schemaLinter(schema)));
+
+      const completionSource = schemaAutocomplete(schema);
+      if (completionSource) {
+        extensions.push(
+          autocompletion({
+            override: [completionSource],
+          }),
+        );
+      }
+    }
+
+    return extensions;
+  }, [field, schema, hideSubmitErrors]);
+
+  return (
+    <ReactCodeMirror
+      value={formControlAttrs.defaultValue ?? ""}
+      ref={editor}
+      extensions={extensions}
+      className={classNames(
+        formControlAttrs.className,
+        "feature-value-editor",
+        {
+          "is-valid": formControlAttrs.isValid,
+          "is-invalid": formControlAttrs.isInvalid,
+        },
```

Comment:
Are these values being set? I don't see the valid or invalid classes getting added to the element at any point.

Reply:
This broke at some point during refactoring, good catch.